### PR TITLE
Sameeran/ff 924 update js sdk common with allocation key

### PR DESCRIPTION
**Pull request by sameerank**
---
labels: mergeable
---
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: [FF-924](https://linear.app/eppo/issue/FF-924/update-js-sdk-common-with-allocation-key)

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

## Description
[//]: # (Describe your changes in detail)

Renames `experimentKey` to `flagKey`
Extracts `allocationKey` from output of `getAssignmentInternal`
Updates `onPostAssignment` to optionally include `allocationKey`
Update `logAssignment` to use `allocationKey` in creation of the event object that goes to the logger (here's where we get the `allocation`, `experiment`, and `featureFlag` event attributes)

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: src/assignment-hooks.ts
Comment (sameerank) [inline src/assignment-hooks.ts:28]: This is, I think, the one decision point this PR where there was some ambiguity.

After assignment, we could either:
1. keep the original `experimentKey` argument and pass in the experiment key  \` ${flagKey}-${allocationKey}\` instead of the just the `flagKey`
2. or change the first arg to `flagKey` argument and pass in the allocation key as an optional argument

I went with option 2. here because 

* It keeps all the original functionality of the `onPostAssignment` function (i.e. not changing the format of the string that gets passed into the function's first argument) and consequently doesn't require an immediate code change
* It also enables custom allocation-level post assignment functionality if necessary, but would require the user to be aware of `allocationKeys` to use this correctly

If I went with 1, the `onPostAssignment` code (customer-end) would need to be updated to handle the new key format, which works against out goal of making this transition to logging `allocationId`s as friction-free as possible

File: src/client/eppo-client.ts
Comment (aarsilv) [inline src/client/eppo-client.ts:103]: `allocationKey` will be `null` for overrides returned by `getAssignmentInternal`, so this is how prevent logging an assignment for an overridden subject. Nice. Howeve,r I think before this change, this SDK was logging overrides 😬

File: src/client/eppo-client.ts
Comment (aarsilv) [inline src/client/eppo-client.ts:246]: Not your code, but I believe by telling TypeScript to treat these things as a type, we could actually end up passing in a numeric and will just be a number stored (and returned from) `boolValue`. It only matters in the edge case of a developer requesting the incorrect variant type for an experiment. But this is a different behavior than our PHP SDK (which would return null) and Java SDK (which would cast)

**Review by aarsilv — APPROVED** [review]
Nice work! Approving with some consistency & cleanup thoughts that you can take or leave as you see fit!

File: src/assignment-hooks.ts
Comment (sameerank) [reply src/assignment-hooks.ts:28]: Ah the options idea is a cool one, but agreed that it's a bit messy

File: src/client/eppo-client.ts
Comment (sameerank) [reply src/client/eppo-client.ts:127]: Yeah I feel like that's basically what we had with the `get_assignment_variation` function in the dynamically typed languages. I also liked the early return logic to prevent logging events. I felt it was more readable than keeping track of nulls and performing actions based on whether something is null. I might try cleaning this up

File: src/client/eppo-client.ts
Comment (sameerank) [reply src/client/eppo-client.ts:103]: Hmmm I'm thinking about this more and now wondering if I should keep the original functionality of logging overrides ..

So there's the `onPreAssignment` function, which is one way to override the assignment

And there's subject overrides (see `getSubjectVariationOverride`) which uses the "overrides" attribute in the RAC, which seems to come from `opted_in_subjects` in the `experiment_variations` table, which I checked and the only company using this is Goldbelly for one specific variation. Also I don't see any UI for supporting this, so this seems to be a pretty custom thing that we built for this one specific purpose.

The `onPreAssignment` call doesn't exist in any of the other SDKs and again there's one variation by one company that's actually using the overrides, so maybe this is all an edge case that doesn't require us to actively work on making all the other SDKs consistent, as far as whether we log or don't log overrides.

BUT I feel like any customer using these features might actually want to see these overridden assignments logged (or how else would they know that the overrides are working ..). So I'm going to change this back to let `null` allocation keys get logged

Comment (sameerank) [reply src/client/eppo-client.ts:103]: Actually nevermind, I don't know what to do with the experiment key when there's no allocation key. I can't leave out the allocation key because we're already doing a `'${flagKey}' or '${flagKey}-${allocationKey}'` to find experiment assignments and putting in any placeholder text for the allocationKey like '${flagKey}-override` is fragile if anyone makes a feature flag key that ends with '-override'

Maybe we can take the stance that overrides are not assignments and so we leave them out of analysis and we leave it to the customers to track this information if they want it, e.g. with the onPostAssignment hook

File: src/client/eppo-client.ts
Comment (sameerank) [reply src/client/eppo-client.ts]: I agree that this behavior is too divergent from the other SDKs and needs more guardrails. I just added a `isExpectedType` method to `EppoValue` like I did in the python and ruby SDKs and return `NullType` if there's a mismatch. I do a `typeof <numeric_variable> === 'number'` comparison. That should suffice, right?

**Pull request merged**
